auth/auth.go

```diff
@@ -0,0 +1,281 @@
+	if config.FailureHandler == nil {
+		return nil, errors.New("missing FailureHandler")
+	} else if config.StateSetter == nil {
+		return nil, errors.New("missing StateSetter")
+	} else if config.StateGetter == nil {
+		return nil, errors.New("missing StateGetter")
+	} else if config.StateDeleter == nil {
+		return nil, errors.New("missing StateDeleter")
+	}
+	return &Handler{config: config}, nil
+}
+
+// LoginHandler returns an HTTP handler that redirects the user to the SAMS
+// authentication page.
+//
+// It passes through the "prompt" and "prompt_auth" query parameters to SAMS.
+func (h *Handler) LoginHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		// Create a new OIDC provider from the issuer URL using OIDC discovery feature.
+		p, err := oidc.NewProvider(ctx, h.config.Issuer)
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "create new provider"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		// Generate and store a random nonce to the session.
+		nonce, err := randomString()
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "generate nonce"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+		err = h.config.StateSetter(w, r, nonce)
+		if err != nil {
+			ctx = WithError(ctx, &Error{
+				StatusCode: http.StatusInternalServerError,
+				Cause:      errors.Wrap(err, "set state"),
+			})
+			h.config.FailureHandler.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		// NOTE: Using same value for both state and nonce is fine because "state" is
+		// for service self-validation (no CSRF) and "nonce" is for IdP validation (no
+		// repeated attempts).
+		params := url.Values{}
+		params.Add("client_id", h.config.ClientID)
+		params.Add("redirect_uri", h.config.RedirectURI)
+		params.Add("state", nonce)
+		params.Add("nonce", nonce)
+		params.Add("response_type", "code")
+		params.Add("scope", strings.Join(scopes.ToStrings(h.config.RequestScopes), " "))
+
+		// Passthrough the IdP-aware query parameters.
+		params.Add("prompt", r.URL.Query().Get("prompt"))
+		params.Add("prompt_auth", r.URL.Query().Get("prompt_auth"))
+
+		http.Redirect(w, r, p.Endpoint().AuthURL+"?"+params.Encode(), http.StatusFound)
+	})
+}
+
+// CallbackHandler returns an HTTP handler that handles the SAMS callback and
+// calls the success handler upon successful authentication. Use
+// UserInfoFromContext to extract the user information.
+func (h *Handler) CallbackHandler(success http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		nonce, err := h.config.StateGetter(r)
+		h.config.StateDeleter(w, r) // Delete the state after getting it to make sure it's one-time use.
```

Comment:
The "... to make sure it's one-time use." would probably be helpful to add as a note in the doc comments elsewhere, since it wasn't immediately obvious (to me at least).

Reply:
🤔 Where to add as a doc comment are you thinking of?

This is really internal detail about how to handle state in the most strict way, as the users of SAMS auth handler, should they care? i.e. All they need to make sure is to implement the `StateStore` interface, and delete state when it is told to delete.